streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,1123 @@
+                        .collect(Collectors.toList());
+
+                final List<KeyValue<Long, Long>> expectedCommittedResult =
+                    computeExpectedResult(committedInputDataDuringFirstUpgrade, committedState);
+                // TODO: if we don't use the custom partitioner, the test hangs here until TX times out and is aborted
+                verifyCommitted(expectedCommittedResult);
+            } else {
+                // retrying TX
+                expectedUncommittedResult.addAll(computeExpectedResult(
+                    uncommittedInputDataBeforeFirstUpgrade
+                        .stream()
+                        .filter(pair -> keyFilterFirstClient.contains(pair.key))
+                        .collect(Collectors.toList()),
+                    new HashMap<>(committedState)
+                ));
+                verifyUncommitted(expectedUncommittedResult);
+
+                errorInjectedClient1.set(false);
+                stateTransitions1.clear();
+                streams1Alpha.close();
+                waitForStateTransition(stateTransitions1, CLOSE_CRASHED);
+            }
+
+            // phase 5: (restart first client)
+            // expected end state per output partition (C == COMMIT; A == ABORT; ---> indicate the changes):
+            //
+            // stop case:
+            //   p-0: 10 rec + C + 5 rec + C
+            //   p-1: 10 rec + C + 5 rec + C
+            //   p-2: 10 rec + C + 5 rec ---> C
+            //   p-3: 10 rec + C + 5 rec ---> C
+            // crash case:
+            //   p-0: 10 rec + C + 4 rec + A + 5 rec ---> C
+            //   p-1: 10 rec + C + 5 rec + A + 5 rec ---> C
+            //   p-2: 10 rec + C + 5 rec ---> C
+            //   p-3: 10 rec + C + 5 rec ---> C
+            commitRequested.set(0);
+            stateTransitions1.clear();
+            stateTransitions2.clear();
+            streams1Beta = getKafkaStreams("appDir1", StreamsConfig.EXACTLY_ONCE_BETA);
+            streams1Beta.setStateListener((newState, oldState) -> stateTransitions1.add(KeyValue.pair(oldState, newState)));
+            streams1Beta.start();
+            waitForStateTransition(stateTransitions1, TWO_REBALANCES_STARTUP, stateTransitions2, TWO_REBALANCES_RUNNING);
+
+            final Set<Long> committedKeys = mkSet(0L, 1L, 2L, 3L);
+            if (!injectError) {
+                committedKeys.removeAll(keyFilterFirstClient);
+            }
+
+            final List<KeyValue<Long, Long>> expectedCommittedResultAfterRestartFirstClient = computeExpectedResult(
+                uncommittedInputDataBeforeFirstUpgrade
+                    .stream()
+                    .filter(pair -> committedKeys.contains(pair.key))
+                    .collect(Collectors.toList()),
+                committedState
+            );
+            verifyCommitted(expectedCommittedResultAfterRestartFirstClient);
+
+            // phase 6: (complete second batch of data; crash: let second client fail on commit)
```

Comment:
This is new: for the crash case, in inject more errors in mixed mode, after we called `sendOffsetsToTransaction()` but before actually committing.